### PR TITLE
Test NodeCode

### PR DESCRIPTION
**Pull request by majocha**
I had a hunch something's not right with `NodeCode`:
https://github.com/dotnet/fsharp/pull/16536#issuecomment-1906467762

I suspect with increased parallelization brought by Transparent Compiler it tends to mangle thread statics it's supposed to preserve (?)

So, this is a failing test.

### CODE REVIEW

**Comment by vzarytovskii** [comment]
FWIW, I am currently getting rid of node code and replacing it with cancellable tasks, thread static preservation is something I've baked into state machine. But it's far from being finished.

**Comment by 0101** [comment]
Good idea to test this. It definitely looks concerning. I'm surprised it didn't cause us any issues this whole time 🤔

File: tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs [hunk 1]
```diff
@@ -211,6 +211,43 @@ let ``Job is restarted if first requestor cancels but keeps running if second re
 type ExpectedException() =
     inherit Exception()
 
+[<Fact>]
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs:214]: Any plan on how to get this into main?
In one of my older PRs, I used this (nunit I think) feature:

```
    [<Test()>]
    [<Ignore("Waiting for test to be aligned", Until = "2023-04-01 12:00:00Z")>]
```

Which means it does not break the CI immediately, but will start breaking it if the problem is ignored for too long.

**Comment by majocha** [comment]
> Good idea to test this. It definitely looks concerning. I'm surprised it didn't cause us any issues this whole time 🤔

I think Threadpool mostly queues continuations on the same thread unless really pushed and our stuff is quite sequential. NodeCode.Parallel was added just recently with the graph-based type checking.

**Comment by 0101** [comment]
Does this reproduce with anything else than `NodeCode.Parallel`? 

And it was actually used before in `IncrementalBuilder`.
https://github.com/dotnet/fsharp/blob/641d0ee179755ba700c3ac780a347f916d740472/src/Compiler/Service/IncrementalBuild.fs#L760

https://github.com/dotnet/fsharp/blob/641d0ee179755ba700c3ac780a347f916d740472/src/Compiler/Service/IncrementalBuild.fs#L829

But maybe those places don't need the thread statics.

File: src/Compiler/Facilities/BuildGraph.fs [hunk 2]
```diff
@@ -193,7 +193,7 @@ type NodeCode private () =
         }
 
     static member Parallel(computations: NodeCode<'T> seq) =
-        computations |> Seq.map (fun (Node x) -> x) |> Async.Parallel |> Node
+        computations |> Seq.map (fun (Node x) -> x) |> Async.Parallel |> wrapThreadStaticInfo |> Node
```
Comment (0101) [inline src/Compiler/Facilities/BuildGraph.fs:196]: @majocha the test passes for me with this change. 

However it doesn't mean that the `computations` will have the same thread statics as the parent computation. Which we could also be relying on somewhere.

Comment (majocha) [reply src/Compiler/Facilities/BuildGraph.fs:196]: Correct, I'll update the test in a moment.

Comment (0101) [reply src/Compiler/Facilities/BuildGraph.fs:196]: That could be probably fixed like this:

```suggestion
        let diagnosticsLogger = DiagnosticsThreadStatics.DiagnosticsLogger
        let phase = DiagnosticsThreadStatics.BuildPhase

        computations
        |> Seq.map (fun (Node x) ->
            async {
                DiagnosticsThreadStatics.DiagnosticsLogger <- diagnosticsLogger
                DiagnosticsThreadStatics.BuildPhase <- phase
                return! x
            })
        |> Async.Parallel
        |> wrapThreadStaticInfo
        |> Node
```

Comment (0101) [reply src/Compiler/Facilities/BuildGraph.fs:196]: But actually looking at the places where it's currently used this is probably not relied on. (Also if it was we would probably have noticed it).

**Comment by vzarytovskii** [comment]
Nothing regarding changes in this code, but all the thread static management becomes more and more convoluted. It's really hard to follow-up what's actually going on.

**Comment by majocha** [comment]
Works now, although I dislike passing state with `finally`. The intent is unclear and it depends on scheduler implementation.

Do we need release notes for this PR?

File: tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs [hunk 3]
```diff
@@ -211,6 +211,44 @@ let ``Job is restarted if first requestor cancels but keeps running if second re
...
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs:214]: Maybe this should go to `FSharp.Compiler.UnitTests/BuildGraphTests.fs` ?

Comment (majocha) [reply tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs:214]: Yep! I'll take care of that.

**Comment by 0101** [comment]
I think we don't need release notes for this. It doesn't affect the users.

**Review by 0101 — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Comment by 0101** [comment]
Hmm, now with this already merged and built into VSIX, I started seeing:

```fsharp
System.AggregateException : One or more errors occurred. ---> Object reference not set to an instance of an object.
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.CheckerExtensions.FSharpChecker-ParseAndCheckDocumentUsingTransparentCompiler@355(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.CheckerExtensions.FSharpChecker-ParseAndCheckDocument@448(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.Document-GetFSharpParseAndCheckResultsAsync@538(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async StartupCode$FSharp-Editor(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.<>c__DisplayClass12_0.<GetRefactoringFromProviderAsync>b__0(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
---> (Inner Exception #0) System.NullReferenceException : Object reference not set to an instance of an object.
   at Internal.Utilities.Collections.Utils.replayDiagnostics@31.Invoke(Tuple`2 arg1)
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action,IEnumerable`1 source)
   at Internal.Utilities.Collections.Utils.replayDiagnostics@31-1.Invoke(IEnumerable`1 source)
   at async StartupCode$FSharp-Compiler-Service[TKey,TVersion,TValue,d](<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async StartupCode$FSharp-Compiler-Service[a,b](<Unknown Parameters>) <--- 
```

I can't quite explain it, even if somehow there was no logger set on a thread, we should still get `AssertFalseDiagnosticsLogger` which will just do nothing. Wonder how it becomes `null` or where the NRE is actually coming from.

**Comment by majocha** [comment]
ResizeArray is not threadsafe?

**Comment by 0101** [comment]
True, but access to it is always under lock, unless I missed something. Also can it throw NRE due to a race condition?

**Comment by majocha** [comment]
Yes, right. There is a lock. We supposedly always initialize the static field to AssertFalseDiagnosticsLogger. That's wierd.

**Comment by majocha** [comment]
So I'd look at places when we call the setter. Lots of them.

**Comment by majocha** [comment]
Now I've seen the exact exception, too, surfacing in VS. But I haven't found a way to reproduce it.

**Comment by 0101** [comment]
Yeah a reliable repro would be nice, but not sure it's even possible. I've tried to get it in debug but couldn't so far.

**Comment by majocha** [comment]
I'm very sure your code is good.
It seems there are some (compiler?) bugs with how DiagnosticsThreadStatics initialization gets optimized?
I tried to make the initialization more straighforward in #16602, let's see how it goes.

File: src/Compiler/Facilities/BuildGraph.fs [hunk 4]
```diff
@@ -193,7 +193,19 @@ type NodeCode private () =
...
 
     static member Parallel(computations: NodeCode<'T> seq) =
-        computations |> Seq.map (fun (Node x) -> x) |> Async.Parallel |> Node
+        let diagnosticsLogger = DiagnosticsThreadStatics.DiagnosticsLogger
+        let phase = DiagnosticsThreadStatics.BuildPhase
+
+        computations
+        |> Seq.map (fun (Node x) ->
+            async {
+                DiagnosticsThreadStatics.DiagnosticsLogger <- diagnosticsLogger
```
Comment (majocha) [inline src/Compiler/Facilities/BuildGraph.fs:202]: This is not safe, as the logger we're passing is most probably not thread-safe. So we're betting on the computation to actually never use it.

**Comment by majocha** [comment]
> Hmm, now with this already merged and built into VSIX, I started seeing:
> 
> ```fsharp
> System.AggregateException : One or more errors occurred. ---> Object reference not set to an instance of an object.
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.CheckerExtensions.FSharpChecker-ParseAndCheckDocumentUsingTransparentCompiler@355(<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.CheckerExtensions.FSharpChecker-ParseAndCheckDocument@448(<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async Microsoft.VisualStudio.FSharp.Editor.WorkspaceExtensions.Document-GetFSharpParseAndCheckResultsAsync@538(<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async StartupCode$FSharp-Editor(<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.<>c__DisplayClass12_0.<GetRefactoringFromProviderAsync>b__0(<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
> ---> (Inner Exception #0) System.NullReferenceException : Object reference not set to an instance of an object.
>    at Internal.Utilities.Collections.Utils.replayDiagnostics@31.Invoke(Tuple`2 arg1)
>    at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action,IEnumerable`1 source)
>    at Internal.Utilities.Collections.Utils.replayDiagnostics@31-1.Invoke(IEnumerable`1 source)
>    at async StartupCode$FSharp-Compiler-Service[TKey,TVersion,TValue,d](<Unknown Parameters>)
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async StartupCode$FSharp-Compiler-Service[a,b](<Unknown Parameters>) <--- 
> ```
> 
> I can't quite explain it, even if somehow there was no logger set on a thread, we should still get `AssertFalseDiagnosticsLogger` which will just do nothing. Wonder how it becomes `null` or where the NRE is actually coming from.

@0101 I suspect we cannot pass a non-thread safe logger here:

https://github.com/dotnet/fsharp/blob/37251d42d51958d75709501f412cd348283d7e36/src/Compiler/Facilities/BuildGraph.fs#L199-L208

We're betting on the computation to actually never use it. I guess in some tasks it does use it causing mayhem.

**Comment by 0101** [comment]
Yeah you're right. We should either make it a thread-safe one or pass a separate one to each computation and then merge the results.

**Comment by majocha** [comment]
> Yeah you're right. We should either make it a thread-safe one or pass a separate one to each computation and then merge the results.

Yes, I think I had something like this in the AsyncLocal branch.

**Comment by majocha** [comment]
@0101 see also #16712 😒
